crates/ruff_linter/src/rules/pylint/rules/unreachable.rs

```diff
@@ -665,6 +665,10 @@ fn post_process_try(
             NextBlock::Terminate => {
                 match block.stmts.last() {
                     Some(Stmt::Return(_)) => {
+                        // if we are already in a `finally` block, terminate
+                        if Some(idx) == finally_index {
+                            return;
+                        }
                         // re-route to finally if present and not already re-routed
                         if let Some(finally_index) = finally_index {
                             blocks.blocks[idx].next = NextBlock::Always(finally_index);
```

Comment:
This would probably be a small refactor and is something better left for its own PR but we could have better runtime assertions around preventing loops by introducing a `blocks.set_next(idx, next)` method that contains an assertion that `idx` and the `next`'s index aren't equal.